k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java

```diff
@@ -0,0 +1,409 @@
+            this.tag = tag;
+            this.commandFactory = commandFactory;
+            this.folder = folder;
+            this.listener = listener;
+        }
+
+        public Builder tag(int tag) {
+            this.tag = tag;
+            return this;
+        }
+
+        public Builder sequenceSet(Collection<Long> sequenceSet) {
+            this.sequenceSet = new HashSet<>(sequenceSet);
+            return this;
+        }
+
+        public Builder addSequenceNumber(Long sequenceNumber) {
+            if (sequenceSet == null) {
+                sequenceSet = new HashSet<>();
+            }
+            sequenceSet.add(sequenceNumber);
+            return this;
+        }
+
+        public Builder sequenceRanges(List<Map.Entry<String, String>> sequenceRanges) {
+            this.sequenceRanges = sequenceRanges;
+            return this;
+        }
+
+        public Builder addSequenceRange(String lowerBound, String upperBound) {
+            if (sequenceRanges == null) {
+                sequenceRanges = new ArrayList<>();
+            }
+            sequenceRanges.add(new AbstractMap.SimpleEntry<String, String>(lowerBound, upperBound));
+            return this;
+        }
+
+        public Builder uidSet(Collection<Long> uidSet) {
+            this.uidSet = new HashSet<>(uidSet);
+            return this;
+        }
+
+        public Builder addUid(Long uid) {
+            if (uidSet == null) {
+                uidSet = new HashSet<>();
+            }
+            uidSet.add(uid);
+            return this;
+        }
+
+        public Builder uidRanges(List<Map.Entry<String, String>> uidRanges) {
+            this.uidRanges = uidRanges;
+            return this;
+        }
+
+        public Builder addUidRange(String lowerBound, String upperBound) {
+            if (uidRanges == null) {
+                uidRanges = new ArrayList<>();
+            }
+            uidRanges.add(new AbstractMap.SimpleEntry<String, String>(lowerBound, upperBound));
+            return this;
+        }
+
+        public Builder queryString(String queryString) {
+            this.queryString = queryString;
+            return this;
+        }
+
+        public Builder performFullTextSearch(boolean performFullTextSearch) {
+            this.performFullTextSearch = performFullTextSearch;
+            return this;
+        }
+
+        public Builder since(Date since) {
+            this.since = since;
+            return this;
+        }
+
+        public Builder requiredFlags(Set<Flag> requiredFlags) {
+            this.requiredFlags = requiredFlags;
+            return this;
+        }
+
+        public Builder forbiddenFlags(Set<Flag> forbiddenFlags) {
+            this.forbiddenFlags = forbiddenFlags;
+            return this;
+        }
+
+        public UidSearchCommand build() {
+            return new UidSearchCommand(this);
+        }
+
+        private int getLength() {
+            return this.build().createCommandString().getBytes().length;
```

Comment:
hmm, this method seems like a hack. looking at its use above, maybe change the loop to a `do .. while();` that builds at the end, checks the length and then uses the result that was built